### ISSUE TITLE
Use globa option 'SDWebImageScaleDownLargeImages'  makes gif imageload crash

### ISSUE THREAD

**Issue by icoder33** [SDWebImage/SDWebImage#3536]
### New Issue Checklist

* [x] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [x] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [x] I have searched for a similar issue in the [project](https://github.com/rs/SDWebImage/issues) and found none

### Issue Info

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform Name           | e.g. ios / macos / tvos / watchos
 Platform Version        | e.g. 12.0 / 10.14.0 / 12.0 / 5.0
 SDWebImage Version      | e.g. 5.0.0 / 4.4.0
 Integration Method      | e.g. carthage / cocoapods / manually
 Xcode Version           | e.g. Xcode 11 / Xcode 10
 Repro rate              | e.g. all the time (100%) / sometimes x% / only once
 Repro with our demo prj | e.g. does it happen with our demo project?
 Demo project link       | e.g. link to a demo project that highlights the issue

### Issue Description and Steps

Please fill in the detailed description of the issue (full output of any stack trace, compiler error, ...) and the steps to reproduce the issue.

If I set SD config like this 
![image](https://github.com/SDWebImage/SDWebImage/assets/11675594/9e3490ef-5725-4f22-950e-f734e9ff215c)
in some  gif image load case, it will crash , here is the stack
![image](https://github.com/SDWebImage/SDWebImage/assets/11675594/8874f612-2de8-4be7-b3a2-7c721f40e5ac)
but if i dont use this option, in some large image load case , it will cause memory issue, for now , our server cant handle it for some reason so I wonder whats the best way to solve or config that

**Comment by dreampiggy** [comment]
1. Which one do you use, `UIImageView` or `SDAnimatedImageView` ? means (`UIAnimatedImage` or `SDAnimatedImage`)
2. Did you try to use `context: @{SDWebImageContextStoreCacheType: @(SDImageCacheTypeMemory)` instead to avoid store the thumbnail one into disk cache (cause `Image Encoding` ) ?

**Comment by icoder33** [comment]
1.For now is UIImageView, but I try SDAnimatedImageView already, its has same issue,
2. not yet , I will try this , but its hard to apply to offical version cause our local web (which use urlschemahandler) need use disk data to render gif or other image  in some case

**Comment by dreampiggy** [comment]
> need use disk data to render gif or other image in some case

In current 5.15.0, when you use `SDWebImageScaleDownLargeImages`, it will actually translate into 

```
@{}](context: @{SDWebImageContextImageThumbnailPixelSize: @(sqrt(SDImageCoderHelper.defaultScaleDownBytes / 4))
```

And

Thumbnail Decode here, by default store `Thumbnailed Image object` into memory cache, Store `re-encoded Thumbnailed Image data` into disk cache (This cause encoding)

The original GIF data is still store into disk cache, and original GIF size image is nil.

In conclusion, when use thumbnail, there are actually 4 thing:

1. Original image data (from network, and store into disk)
2. Original image object (nil, because thumbnail not create full image object)
3. Thumbnailed image data (nil, which cause SDImageCache to re-encode the image object into data, `Image Encoding`)
4. Thumbnailed image object (decoded from original data and thumbnail size, and store into memory)

**Comment by icoder33** [comment]
okey thanks for your apply again , understand ~

**Comment by dreampiggy** [comment]
To quickly test, use #3537  branch for testing.

Remove your old `imageThumbnailPixelSize` and you will see the GIF frame pixel size become smaller

You can control the limit bytes using context-level control `SDWebImageContextImageScaleDownLimitBytes`  (bytes size)

Or use the global control via `SDImageCoderHelper.defaultScaleDownLimitBytes`

Provide your feedback for this PR

**Comment by dreampiggy** [comment]
In SDWebImage demo, this URL: https://raw.githubusercontent.com/liyong03/YLGIFImage/master/YLGIFImageDemo/YLGIFImageDemo/joy.gif

The original pixel size is 800x600

And the default 40MB limits will cause this into

```
(CGSize) size = (width = 232, height = 174)
```

**Comment by icoder33** [comment]
Okey ~ Thanks For Your Help

**Comment by dreampiggy** [comment]
> To quickly test, use #3537 branch for testing.
> 
> Replace your old `imageThumbnailPixelSize` with new option (or just `SDWebImageScaleDownLargeImages` with default limit) and you will see the GIF frame pixel size become smaller
> 
> You can control the limit bytes using context-level control `SDWebImageContextImageScaleDownLimitBytes` (bytes size)
> 
> Or use the global control via `SDImageCoderHelper.defaultScaleDownLimitBytes`
> 
> Provide your feedback for this PR

Did you have time to test this and check result ?

**Comment by icoder33** [comment]
okey I will try later

**Comment by dreampiggy** [comment]
Seems too many time no reply. I'll merge the #3537 and release 5.16.0

**Comment by dreampiggy** [comment]
Test 5.16.0 and report more issue or re-open this:

https://github.com/SDWebImage/SDWebImage/releases/tag/5.16.0
